miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/util/MiniAppDatabaseEncryptionUtil.kt

```diff
@@ -0,0 +1,169 @@
+package com.rakuten.tech.mobile.miniapp.storage.util
+
+import android.content.Context
+import android.os.Build
+import android.util.Base64
+import com.google.gson.Gson
+import com.google.gson.JsonSyntaxException
+import com.google.gson.reflect.TypeToken
+import java.security.SecureRandom
+import java.security.spec.KeySpec
+import javax.crypto.Cipher
+import javax.crypto.SecretKey
+import javax.crypto.SecretKeyFactory
+import javax.crypto.spec.IvParameterSpec
+import javax.crypto.spec.PBEKeySpec
+import javax.crypto.spec.SecretKeySpec
+
+private const val ENCRYPTION_ALGORITHM = "AES"
+private const val SHARED_PREFERENCE_KEY = "PASSCODE"
+private const val SHARED_PREFERENCE_NAME = "MiniAppDatabase"
+private const val CIPHER_TRANSFORMATION = "AES/CBC/PKCS5Padding"
+private const val SECRET_KEY_ALGORITHM = "PBKDF2WithHmacSHA256"
+
+/**
+ * Container for everything needed for decrypting the database.
+ *
+ * @param iv initialization vector
+ * @param salt cryptographic salt
+ * @param encryptedPasscode encrypted database key
+ */
+private data class EncryptedPasscodeDataHolder(
+    val iv: String,
+    val salt: String,
+    val encryptedPasscode: String
+)
+
+private const val PASSCODE_DECRYPTION_ERROR = "Failed to decrypt database key"
+
+/**
+ * Singleton Database Encryption Util Class
+ * Using AES/CBC/PKCS5Padding algorithm to encrypt the
+ * database passcode and the encrypted passcode will be
+ * stored to the preferences to pass the same encrypted
+ * passcode lock everytime to lock the access of the
+ * database from any third party database browser app.
+ *
+ * If needed in the future then this passcode can be taken from
+ * the user with a enter passcode UI screen to access the Database.
+ */
+object MiniAppDatabaseEncryptionUtil {
```

Comment:
I think, you didn't want to expose it to HostApp, so we can define `internal object` here.